merlion/models/layers.py

```diff
@@ -0,0 +1,253 @@
+#
+"""
+Base class for _layered_ models. These are models which act as a wrapper around another model, often with additional
+functionality. This is the basis for :py:mod:`default models <merlion.models.defaults>` and
+:py:mod:`AutoML models <merlion.models.automl>`_.
+"""
+from copy import deepcopy
+from typing import Any, Dict, Union
+
+from merlion.models.base import Config, ModelBase
+from merlion.models.factory import ModelFactory
+from merlion.models.anomaly.base import DetectorBase, DetectorConfig
+from merlion.models.forecast.base import ForecasterBase, ForecasterConfig
+from merlion.models.anomaly.forecast_based.base import ForecastingDetectorBase
+from merlion.utils import TimeSeries
+from merlion.utils.misc import AutodocABCMeta
+
+
+class LayeredModelConfig(Config):
+    """
+    Config object for a `LayeredModel`. See `LayeredModel` documentation for more details.
+    """
+
+    def __init__(self, model: Union[ModelBase, Dict], model_kwargs=None, **kwargs):
+        """
+        :param model: The model being wrapped, or a dict representing it.
+        :param model_kwargs: Keyword arguments used specifically to initialize the underlying model. Only used if
+            ``model`` is a dict. Will override keys in the ``model`` dict if specified.
+        :param kwargs: Any other keyword arguments (e.g. for initializing a base class). If ``model`` is a dict,
+            we will also try to pass these arguments when creating the actual underlying model. However, they will
+            not override arguments in either the ``model`` dict or ``model_kwargs`` dict.
+        """
+        # Model-specific kwargs override kwargs when creating the model.
+        model_kwargs = {} if model_kwargs is None else model_kwargs
+        if isinstance(model, dict):
+            model.update({k: v for k, v in kwargs.items() if k not in model and k not in model_kwargs})
+            model, extra_kwargs = ModelFactory.create(**{**model, **model_kwargs, "return_unused_kwargs": True})
+            kwargs.update(extra_kwargs)
+        self.model = model
+        super().__init__(**kwargs)
+
+        # If no model was created, reserve unused kwargs to try initializing the model with
+        if model is None:
+            extra_kwargs = {k: v for k, v in kwargs.items() if k not in self.to_dict()}
+            self.model_kwargs = {**extra_kwargs, **model_kwargs}
+
+    @property
+    def model(self):
+        """
+        The underlying model which this layer wraps.
+        """
+        return self._model
+
+    @model.setter
+    def model(self, model):
+        self._model = model
+        self.model_kwargs = {}
+
+    @property
+    def depth(self):
+        """
+        The depth of the layered model, i.e. the number of layers until we hit a base model. Base models have depth 0.
+        """
+        return 0 if self.model is None else self.model.config.depth + 1
```

Comment:
Having a manual `config.depth` field seems strange to me; I haven't really found a lot of nested structures work like this. At best we can figure out the depth from the config that's passed in (by analyzing the depth of the JSON)?

Reply:
I'll remove the depth field. I'll discuss the linked list vs. tree structure in more detail in your other comment.